qemu/tests/block_hotplug.py

```diff
@@ -24,48 +25,85 @@ def run(test, params, env):
     :param params: Dictionary with the test parameters.
     :param env:    Dictionary with test environment.
     """
-    def find_disk(vm, cmd):
-        """
-        Find all disks in guest.
-        """
-        if params.get("os_type") == "linux":
-            pattern = params.get("get_disk_pattern", "^/dev/vd[a-z]*$")
-        elif params.get("os_type") == "windows":
-            pattern = r"^\d+"
-            cmd = params.get("get_disk_index", "wmic diskdrive get index")
+    def _find_all_disks(session):
+        """ Find all disks in guest. """
+        global all_disks
+        if windows:
+            all_disks = set(session.cmd('wmic diskdrive get index').split()[1:])
         else:
-            test.cancel("Unsupported OS type '%s'" % params.get("os_type"))
-
+            all_disks = utils_misc.list_linux_guest_disks(session)
+        return all_disks
+
+    def run_sub_test(test_name):
+        """ Run subtest before/after hotplug/unplug device. """
+        error_context.context("Running sub test '%s'." % test_name, logging.info)
+        utils_test.run_virt_sub_test(test, params, env, test_name)
+
+    def wait_plug_disks(session, action, disks_before_plug, excepted_num):
+        """ Wait plug disks completely. """
+        if not utils_misc.wait_for(lambda: len(disks_before_plug ^ _find_all_disks(
+                session)) == excepted_num, 30, step=1.5):
```

Comment:
1. suggest to set timeout to 60s
2. suggest to list disk information before raise error
  for windows: wmic logicaldisk get drivetype,name,description
                       wmic diskdrive list brief /format:list
  for linux: lsblk

Reply:
OK, will update it.
Thanks.